headless mode not working on Linux

examples/headless can't compile

```
gcc headless.c ../../tigr.c -Os -o headless -fsanitize=undefined -I../.. -Wall -DTIGR_HEADLESS -s -lGLU -lGL -lX11
../../tigr.c:114:5: error: unknown type name ‘GLXContext’
  114 |     GLXContext glc;
      |     ^~~~~~~~~~
make: *** [Makefile:14: headless] Error 1
```

--- a/ci/Makefile
+++ b/ci/Makefile
@@ -1,13 +1,13 @@
-CFLAGS = -I..
+CFLAGS += -I..
 ifeq ($(OS),Windows_NT)
-	LDFLAGS = -s -lopengl32 -lgdi32
+	LDFLAGS += -s -lopengl32 -lgdi32
 	EXT = .exe
 else
 	UNAME_S := $(shell uname -s)
 	ifeq ($(UNAME_S),Darwin)
-		LDFLAGS = -framework OpenGL -framework Cocoa
+		LDFLAGS += -framework OpenGL -framework Cocoa
 	else ifeq ($(UNAME_S),Linux)
-		LDFLAGS = -s -lGLU -lGL -lX11
+		LDFLAGS += -s -lGLU -lGL -lX11
 	endif
 endif
 

--- a/examples/.gitignore
+++ b/examples/.gitignore
@@ -1,6 +1,10 @@
-shader/shader
-opengl/opengl
-hello/hello
+clip/clip
 demo/demo
+flags/flags
+headless/headless
+hello/hello
+opengl/opengl
+shader/shader
+
 *.o
 *.exe

--- a/examples/.gitignore
+++ b/examples/.gitignore
@@ -2,6 +2,7 @@ clip/clip
 demo/demo
 flags/flags
 headless/headless
+headless/headless.png
 hello/hello
 opengl/opengl
 shader/shader

--- a/examples/clip/Makefile
+++ b/examples/clip/Makefile
@@ -1,12 +1,12 @@
-CFLAGS = -I../.. -Os
+CFLAGS += -I../.. -Os
 ifeq ($(OS),Windows_NT)
-	LDFLAGS = -s -lopengl32 -lgdi32
+	LDFLAGS += -s -lopengl32 -lgdi32
 else
 	UNAME_S := $(shell uname -s)
 	ifeq ($(UNAME_S),Darwin)
-		LDFLAGS = -framework OpenGL -framework Cocoa
+		LDFLAGS += -framework OpenGL -framework Cocoa
 	else ifeq ($(UNAME_S),Linux)
-		LDFLAGS = -s -lGLU -lGL -lX11
+		LDFLAGS += -s -lGLU -lGL -lX11
 	endif
 endif
 

--- a/examples/demo/Makefile
+++ b/examples/demo/Makefile
@@ -1,12 +1,12 @@
-CFLAGS = -I../..
+CFLAGS += -I../..
 ifeq ($(OS),Windows_NT)
-	LDFLAGS = -s -lopengl32 -lgdi32
+	LDFLAGS += -s -lopengl32 -lgdi32
 else
 	UNAME_S := $(shell uname -s)
 	ifeq ($(UNAME_S),Darwin)
-		LDFLAGS = -framework OpenGL -framework Cocoa
+		LDFLAGS += -framework OpenGL -framework Cocoa
 	else ifeq ($(UNAME_S),Linux)
-		LDFLAGS = -s -lGLU -lGL -lX11
+		LDFLAGS += -s -lGLU -lGL -lX11
 	endif
 endif
 

--- a/examples/flags/Makefile
+++ b/examples/flags/Makefile
@@ -1,12 +1,12 @@
-CFLAGS = -I../..
+CFLAGS += -I../..
 ifeq ($(OS),Windows_NT)
-	LDFLAGS = -s -lopengl32 -lgdi32
+	LDFLAGS += -s -lopengl32 -lgdi32
 else
 	UNAME_S := $(shell uname -s)
 	ifeq ($(UNAME_S),Darwin)
-		LDFLAGS = -framework OpenGL -framework Cocoa
+		LDFLAGS += -framework OpenGL -framework Cocoa
 	else ifeq ($(UNAME_S),Linux)
-		LDFLAGS = -s -lGLU -lGL -lX11
+		LDFLAGS += -s -lGLU -lGL -lX11
 	endif
 endif
 

--- a/examples/headless/Makefile
+++ b/examples/headless/Makefile
@@ -1,4 +1,4 @@
-CFLAGS = -I../.. -Wall -DTIGR_HEADLESS
+CFLAGS += -I../.. -Wall -DTIGR_HEADLESS
 
 headless : headless.c ../../tigr.c
 	gcc $^ -Os -o $@ $(CFLAGS)

--- a/examples/hello/Makefile
+++ b/examples/hello/Makefile
@@ -1,12 +1,12 @@
-CFLAGS = -I../..
+CFLAGS += -I../..
 ifeq ($(OS),Windows_NT)
-	LDFLAGS = -s -lopengl32 -lgdi32
+	LDFLAGS += -s -lopengl32 -lgdi32
 else
 	UNAME_S := $(shell uname -s)
 	ifeq ($(UNAME_S),Darwin)
-		LDFLAGS = -framework OpenGL -framework Cocoa
+		LDFLAGS += -framework OpenGL -framework Cocoa
 	else ifeq ($(UNAME_S),Linux)
-		LDFLAGS = -s -lGLU -lGL -lX11
+		LDFLAGS += -s -lGLU -lGL -lX11
 	endif
 endif
 

--- a/examples/opengl/Makefile
+++ b/examples/opengl/Makefile
@@ -1,12 +1,12 @@
-CFLAGS = -I../..
+CFLAGS += -I../..
 ifeq ($(OS),Windows_NT)
-	LDFLAGS = -s -lopengl32 -lgdi32
+	LDFLAGS += -s -lopengl32 -lgdi32
 else
 	UNAME_S := $(shell uname -s)
 	ifeq ($(UNAME_S),Darwin)
-		LDFLAGS = -framework OpenGL -framework Cocoa
+		LDFLAGS += -framework OpenGL -framework Cocoa
 	else
-		LDFLAGS = -s -lGLU -lGL -lX11
+		LDFLAGS += -s -lGLU -lGL -lX11
 	endif
 endif
 

--- a/examples/shader/Makefile
+++ b/examples/shader/Makefile
@@ -1,12 +1,12 @@
-CFLAGS = -I../..
+CFLAGS += -I../..
 ifeq ($(OS),Windows_NT)
-	LDFLAGS = -s -lopengl32 -lgdi32
+	LDFLAGS += -s -lopengl32 -lgdi32
 else
 	UNAME_S := $(shell uname -s)
 	ifeq ($(UNAME_S),Darwin)
-		LDFLAGS = -framework OpenGL -framework Cocoa
+		LDFLAGS += -framework OpenGL -framework Cocoa
 	else ifeq ($(UNAME_S),Linux)
-		LDFLAGS = -s -lGLU -lGL -lX11
+		LDFLAGS += -s -lGLU -lGL -lX11
 	endif
 endif
 

--- a/src/tigr_inflate.c
+++ b/src/tigr_inflate.c
@@ -66,7 +66,7 @@ static void copy(State* s, const unsigned char* src, int len) {
 }
 
 static int build(State* s, unsigned* tree, unsigned char* lens, int symcount) {
-    int n, codes[16], first[16], counts[16] = { 0 };
+    unsigned n, codes[16], first[16], counts[16] = { 0 };
 
     // Frequency count.
     for (n = 0; n < symcount; n++)
@@ -84,7 +84,7 @@ static int build(State* s, unsigned* tree, unsigned char* lens, int symcount) {
     for (n = 0; n < symcount; n++) {
         int len = lens[n];
         if (len != 0) {
-            int code = codes[len]++, slot = first[len]++;
+            unsigned code = codes[len]++, slot = first[len]++;
             tree[slot] = (code << (32 - len)) | (n << 4) | len;
         }
     }

--- a/src/tigr_internal.h
+++ b/src/tigr_internal.h
@@ -30,7 +30,7 @@ void tigrPosition(Tigr* bmp, int scale, int windowW, int windowH, int out[4]);
 #include <windows.h>
 #endif
 
-#if __linux__ && !__ANDROID__
+#if !defined(TIGR_HEADLESS) && __linux__ && !__ANDROID__
 #include <X11/X.h>
 #include <X11/Xlib.h>
 #endif

--- a/src/tigr_internal.h
+++ b/src/tigr_internal.h
@@ -91,8 +91,6 @@ typedef struct {
     int shown, closed;
 #ifdef TIGR_GAPI_GL
     GLStuff gl;
-#endif
-
 #ifdef _WIN32
     wchar_t* wtitle;
     DWORD dwStyle;
@@ -108,6 +106,7 @@ typedef struct {
     XIC ic;
 #endif  // __ANDROID__
 #endif  // __linux__
+#endif  // TIGR_GAPI_GL
 
     Tigr* widgets;
     int widgetsWanted;

--- a/src/tigr_linux.c
+++ b/src/tigr_linux.c
@@ -6,6 +6,7 @@
 
 #include <stdio.h>
 #include <stdlib.h>
+#include <stdbool.h>
 #include <string.h>
 #include <stdarg.h>
 #include <sys/time.h>
@@ -13,6 +14,7 @@
 #include <X11/Xlib.h>
 #include <X11/Xlocale.h>
 #include <X11/XKBlib.h>
+#include <X11/Xatom.h>
 #include <GL/glx.h>
 
 static Display* dpy;
@@ -173,20 +175,10 @@ Tigr* tigrWindow(int w, int h, const char* title, int flags) {
     XMapWindow(dpy, xwin);
 
     if (flags & TIGR_FULLSCREEN) {
-        // https://www.tonyobryan.com//index.php?article=9
-        WindowHints hints;
-        Atom property;
-        hints.flags = 2;
-        hints.decorations = 0;
-        property = XInternAtom(dpy, "_MOTIF_WM_HINTS", True);
-        XChangeProperty(dpy, xwin, property, property, 32, PropModeReplace, (unsigned char*)&hints, 5);
-        int screen = DefaultScreen(dpy);
-        int dWidth = DisplayWidth(dpy, screen);
-        int dHeight = DisplayHeight(dpy, screen);
-        XMoveResizeWindow(dpy, xwin, 0, 0, dWidth, dHeight);
-        XMapRaised(dpy, xwin);
-        XGrabPointer(dpy, xwin, True, 0, GrabModeAsync, GrabModeAsync, xwin, 0L, CurrentTime);
-        XGrabKeyboard(dpy, xwin, False, GrabModeAsync, GrabModeAsync, CurrentTime);
+        // https://superuser.com/questions/1680077/does-x11-actually-have-a-native-fullscreen-mode
+        Atom wm_state   = XInternAtom (dpy, "_NET_WM_STATE", true );
+        Atom wm_fullscreen = XInternAtom (dpy, "_NET_WM_STATE_FULLSCREEN", true );
+        XChangeProperty(dpy, xwin, wm_state, XA_ATOM, 32, PropModeReplace, (unsigned char *)&wm_fullscreen, 1);
     } else {
         // Wait for window to get mapped
         for (;;) {

--- a/tigr.c
+++ b/tigr.c
@@ -37,7 +37,7 @@ void tigrPosition(Tigr* bmp, int scale, int windowW, int windowH, int out[4]);
 #include <windows.h>
 #endif
 
-#if __linux__ && !__ANDROID__
+#if !defined(TIGR_HEADLESS) && __linux__ && !__ANDROID__
 #include <X11/X.h>
 #include <X11/Xlib.h>
 #endif

--- a/tigr.c
+++ b/tigr.c
@@ -4358,6 +4358,7 @@ void tigr_android_destroy();
 
 #include <stdio.h>
 #include <stdlib.h>
+#include <stdbool.h>
 #include <string.h>
 #include <stdarg.h>
 #include <sys/time.h>
@@ -4365,6 +4366,7 @@ void tigr_android_destroy();
 #include <X11/Xlib.h>
 #include <X11/Xlocale.h>
 #include <X11/XKBlib.h>
+#include <X11/Xatom.h>
 #include <GL/glx.h>
 
 static Display* dpy;
@@ -4525,20 +4527,10 @@ Tigr* tigrWindow(int w, int h, const char* title, int flags) {
     XMapWindow(dpy, xwin);
 
     if (flags & TIGR_FULLSCREEN) {
-        // https://www.tonyobryan.com//index.php?article=9
-        WindowHints hints;
-        Atom property;
-        hints.flags = 2;
-        hints.decorations = 0;
-        property = XInternAtom(dpy, "_MOTIF_WM_HINTS", True);
-        XChangeProperty(dpy, xwin, property, property, 32, PropModeReplace, (unsigned char*)&hints, 5);
-        int screen = DefaultScreen(dpy);
-        int dWidth = DisplayWidth(dpy, screen);
-        int dHeight = DisplayHeight(dpy, screen);
-        XMoveResizeWindow(dpy, xwin, 0, 0, dWidth, dHeight);
-        XMapRaised(dpy, xwin);
-        XGrabPointer(dpy, xwin, True, 0, GrabModeAsync, GrabModeAsync, xwin, 0L, CurrentTime);
-        XGrabKeyboard(dpy, xwin, False, GrabModeAsync, GrabModeAsync, CurrentTime);
+        // https://superuser.com/questions/1680077/does-x11-actually-have-a-native-fullscreen-mode
+        Atom wm_state   = XInternAtom (dpy, "_NET_WM_STATE", true );
+        Atom wm_fullscreen = XInternAtom (dpy, "_NET_WM_STATE_FULLSCREEN", true );
+        XChangeProperty(dpy, xwin, wm_state, XA_ATOM, 32, PropModeReplace, (unsigned char *)&wm_fullscreen, 1);
     } else {
         // Wait for window to get mapped
         for (;;) {

--- a/tigr.c
+++ b/tigr.c
@@ -1212,7 +1212,7 @@ static void copy(State* s, const unsigned char* src, int len) {
 }
 
 static int build(State* s, unsigned* tree, unsigned char* lens, int symcount) {
-    int n, codes[16], first[16], counts[16] = { 0 };
+    unsigned n, codes[16], first[16], counts[16] = { 0 };
 
     // Frequency count.
     for (n = 0; n < symcount; n++)
@@ -1230,7 +1230,7 @@ static int build(State* s, unsigned* tree, unsigned char* lens, int symcount) {
     for (n = 0; n < symcount; n++) {
         int len = lens[n];
         if (len != 0) {
-            int code = codes[len]++, slot = first[len]++;
+            unsigned code = codes[len]++, slot = first[len]++;
             tree[slot] = (code << (32 - len)) | (n << 4) | len;
         }
     }

--- a/tigr.c
+++ b/tigr.c
@@ -98,8 +98,6 @@ typedef struct {
     int shown, closed;
 #ifdef TIGR_GAPI_GL
     GLStuff gl;
-#endif
-
 #ifdef _WIN32
     wchar_t* wtitle;
     DWORD dwStyle;
@@ -115,6 +113,7 @@ typedef struct {
     XIC ic;
 #endif  // __ANDROID__
 #endif  // __linux__
+#endif  // TIGR_GAPI_GL
 
     Tigr* widgets;
     int widgetsWanted;

